feat: Use perfect hash tables for file types and icons

cfxegbert opened the pull request:
Stripped down version of pull request https://github.com/ogham/exa/pull/1229.  I will add other new icons and add icon constants in other pull requests.

This should match the filetypes and icons on the current eza master branch with the following changes:
* Icon added for .ninja files.  Ninja files were already in the filetype table but did not have an icon.
* Icon for .dmg changed from zip icon to disk image icon

cafkafk reviewed (changes requested):
This looks really cool, but since it's such a substantial PR, I won't be able to review it fully as fast as the more trivial PRs.

Currently, it needs to be checked with clippy to pass CI, see https://github.com/eza-community/eza/actions/runs/5996553612/job/16263898274?pr=173

cfxegbert commented:
> This looks really cool, but since it's such a substantial PR, I won't be able to review it fully as fast as the more trivial PRs.

Sorry about the large change.  It is mostly pulling various tables in the code into a few tables in the `build.rs` file.

> Currently, it needs to be checked with clippy to pass CI, see https://github.com/eza-community/eza/actions/runs/5996553612/job/16263898274?pr=173

These are compile time generated files.  I have suppressed the clippy check `unreadable_literal` on the generated tables.

cfxegbert commented:
I just realized I can use the macro version of phf instead of the code generation version.  This would allow me to move the tables out of `build.rs` and back into `icons.rs` and `filetype.rs`.  This would make it easier for other people looking at adding icons/filetypes to find the tables.  Should I go ahead and do this refactor?

cafkafk commented:
> This would make it easier for other people looking at adding icons/filetypes to find the tables.

Then yes, that sounds very good. Icons changes seem very popular.

cfxegbert commented:
Added constants for most of the commonly used icons.  In the process combined nf-dev-windows (e70f) and nf-fa-windows (f17a)

cfxegbert commented:
Added constants for the rest of the icons used multiple times.
* Changed epub icon nf-fae-book_open (e28a) to match ebook and mobi icon nf-fae-book_open_o (e28b)
* Changed .vscode icon from nf-dev-visualstudio (e70c) to nf-md-microsoft_visual_studio_code (f0a1e)
* Changed rtf icon nf-md-file_document (f0219) to match rst and txt icon nf-fa-file_text (f15c)

cafkafk commented on src/info/filetype.rs:
```diff
@@ -8,310 +8,273 @@
-        ]) {
-            true
+impl FileExtensions {
+    fn get_file_type(file: &File<'_>) -> Option<FileType> {
```
This should have a doc comment

cafkafk commented on src/output/icons.rs:
```diff
@@ -45,394 +500,18 @@ pub fn iconify_style(style: Style) -> Style {
-    };
-}
-
 pub fn icon_for_file(file: &File<'_>) -> char {
```
This should have a doc comment, it's fine with just a single line.

cafkafk reviewed (changes requested):
```
Benchmark 1: eza --blocksize -l --hyperlink --grid --git-repos -x --time-style relative ../ --git --header --icons
  Time (mean ± σ):      1.840 s ±  0.082 s    [User: 3.204 s, System: 2.419 s]
  Range (min … max):    1.765 s …  2.033 s    10 runs
 
Benchmark 2: ./target/release/eza --blocksize -l --hyperlink --grid --git-repos -x --time-style relative ../ --git --header --icons
  Time (mean ± σ):      1.787 s ±  0.031 s    [User: 3.145 s, System: 2.413 s]
  Range (min … max):    1.736 s …  1.828 s    10 runs
 
Summary
  ./target/release/eza --blocksize -l --hyperlink --grid --git-repos -x --time-style relative ../ --git --header --icons ran
    1.03 ± 0.05 times faster than eza --blocksize -l --hyperlink --grid --git-repos -x --time-style relative ../ --git --header --icons
```
I haven't reviewed the actual changes to icons yet.

cafkafk reviewed (approved):
I think this is ready, seems to work as expected

cafkafk commented:
I had to interactive rebase to get commit summaries to conform to conventional commits:
```
FAIL  3a18a6c  wrong type: doc  doc: Add function documentation for get_...
FAIL  23c6432  wrong type: task  task: Add constants for the rest of icon...
FAIL  84239f1  wrong type: task  task: Add constants for most of the comm...

3/10 failed
```

But in the process, it seems github no longer sees your authorship, even with your set as a coauthor. 

Perhaps you wanna do an interactive rebase on your end and the force push it? Just change to commit summaries like I did and it should be fine.

If not this should be ready to merge, just let me know.

cfxegbert commented:
I have rebased on main and updated the git log messages.
